lms/djangoapps/instructor/permissions.py

```diff
@@ -32,6 +32,21 @@
 VIEW_ENROLLMENTS = 'instructor.view_enrollments'
 VIEW_FORUM_MEMBERS = 'instructor.view_forum_members'
 
+# Due to how the roles iheritance is implemented currently, eshe_instructor and teaching_assistant have implicit
+# staff access, but unlike staff, they shouldn't be able to enroll and do grade-related operations as per client's
+# requirements. At the same time, all other staff-derived roles, like Limited Staff, should be able to enroll students.
+# This solution is far from perfect, but it's probably the best we can do untill the roles system is reworked.
+_is_teaching_assistant = HasRolesRule('teaching_assistant')
+_is_eshe_instructor = HasRolesRule('eshe_instructor')
+_is_eshe_instructor_or_teaching_assistant = _is_teaching_assistant | _is_eshe_instructor
+is_staff_but_not_teaching_assistant = (
+    (_is_teaching_assistant & HasAccessRule('staff', strict=True)) |
+    (~_is_teaching_assistant & HasAccessRule('staff'))
+)
+is_staff_but_not_eshe_instructor_or_teaching_assistant = (
+    (_is_eshe_instructor_or_teaching_assistant & HasAccessRule('staff', strict=True)) |
+    (~_is_eshe_instructor_or_teaching_assistant & HasAccessRule('staff'))
+)
```

Comment:
What are the other roles with implicit staff access? Would it make sense to have `HasAccessRule('staff', strict=True) & HasRolesRule('implicit_staff_role', strict=True) ` ?

Reply:
@CefBoud, for now we have only one such role: [Limited Staff](https://github.com/openedx/edx-platform/blob/63d18480605e29d19ce89f8d671d7e1a2e217dd2/common/djangoapps/student/roles.py#L297-L302).

Yeah, this will work (I think we don't need `strict=True` for derived roles):
```python
perms[OVERRIDE_GRADES] = HasRolesRule('eshe_instructor') & HasRolesRule('limited_staff') & HasAccessRule('staff', strict=True)
```

I agree that it looks much cleaner. However, I worry about someone creating a new role based on Staff, forgetting to add `HasRolesRule('new_role') & ...` here and thus unknowingly introducing a bug with non-working grades overriding for this new role.

2U de-prioritized permissions refactoring, so we'll have to keep this hack for a while (or forever) 😞, and therefore we need to ensure somehow no one shoot oneself in the foot.

Reply:
Actually, after giving this a second though, I'm not 100% what I described is what you're proposing, @CefBoud. Could you elaborate?

Reply:
@0x29a what you described is exactly what I was proposing.
It indeed looks cleaner to me but, as you mentioned, there is the risk of someone adding a new role with implicit staff permissions and forgetting to amend accordingly. 
So if you think it is better to leave it as is, it's fine by me.